File: packages/core/src/video-resizer/video-resizer.ts
import { exec } from "child_process";
// medium : 960 X 540 export

export type VideoResolution =
  | "ULTRA_LOW"
  | "LOW"
  | "MEDIUM"
  | "HIGH"
  | "ULTRA_HIGH";

const resolutionMap: Record<
  VideoResolution,
  { width: number; height: number }
> = {
  ULTRA_HIGH: {
    height: 1920,
    width: 1080,
  },
  HIGH: {
    height: 1280,
    width: 720,
  },
  MEDIUM: {
    height: 960,
    width: 540,
  },
  LOW: {
    height: 640,
    width: 360,
  },
  ULTRA_LOW: {
    height: 320,
    width: 180,
  },
};

type ResizeVideoInput = {
  videoPath: string;
  resolution: VideoResolution;
  outputFilePath: string;
};

export const resizeVideo = async (input: ResizeVideoInput) => {
  return new Promise<void>(async (resolve, reject) => {
    try {
      const { outputFilePath, resolution, videoPath } = input;
      const dimenstions = resolutionMap[resolution];
      const command = `ffmpeg -y -i ${videoPath} -vf "scale=${dimenstions.width}:${dimenstions.height}" -c:a copy ${outputFilePath}`;
      const cp = exec(command, (err, stdout) => {
        if (err) {
          console.error(err);
          throw err;
        }

        console.log(stdout);
      });

      cp.on("close", () => {
        resolve();
      });
    } catch (error) {
      reject(error);
    }
  });
};

File: server-workspace/core/video-resizer.go
package core

import (
	"errors"
	"os/exec"
	"strconv"

	util "quick-reel.com/util"
)

type Resolution string

const (
	ULTRA_HD_2160p       Resolution = "ULTRA_HD_2160p"
	QUAD_HD_1440p        Resolution = "QUAD_HD_1440p"
	FULL_HD_1080p        Resolution = "FULL_HD_1080p"
	HD_720p              Resolution = "HD_720p"
	LOW_RES_HD_540p      Resolution = "LOW_RES_HD_540p"
	SD_480p              Resolution = "SD_480p"
	VERY_LOW_SD_360p     Resolution = "VERY_LOW_SD_360p"
	EXTREMLY_LOW_SD_240p Resolution = "EXTREMLY_LOW_SD_240p"
	ULTRA_LOW_SD_180p    Resolution = "ULTRA_LOW_SD_180p"
	MINIMAL_SD_144p      Resolution = "MINIMAL_SD_144p"
	LOWEST_SD_120p       Resolution = "LOWEST_SD_120p"
	BARE_MINIMUM_SD_96p  Resolution = "BARE_MINIMUM_SD_96p"
)

type Dimenstions struct {
	Width  int
	Height int
}

var resolutions = map[Resolution]Dimenstions{
	"ULTRA_HD_2160p": {
		Height: 3840,
		Width:  2160,
	},
	"QUAD_HD_1440p": {
		Height: 2560,
		Width:  1440,
	},
	"FULL_HD_1080p": {
		Height: 1920,
		Width:  1080,
	},
	"HD_720p": {
		Height: 1280,
		Width:  720,
	},
	"LOW_RES_HD_540p": {
		Height: 960,
		Width:  540,
	},
	"SD_480p": {
		Height: 854,
		Width:  480,
	},
	"VERY_LOW_SD_360p": {
		Height: 640,
		Width:  360,
	},
	"EXTREMLY_LOW_SD_240p": {
		Height: 426,
		Width:  240,
	},
	"ULTRA_LOW_SD_180p": {
		Height: 320,
		Width:  180,
	},
	"MINIMAL_SD_144p": {
		Height: 256,
		Width:  144,
	},
	"LOWEST_SD_120p": {
		Height: 213,
		Width:  120,
	},
	"BARE_MINIMUM_SD_96p": {
		Height: 170,
		Width:  96,
	},
}

type ResizeVideoInput struct {
	VideoPath  string
	Resolution Resolution
	OutputPath string
}

func ResizeVideo(input ResizeVideoInput) error {
	dimensions, found := resolutions[input.Resolution]
	if !found {
		return errors.New("resolution not found")
	}

	cmd := exec.Command("ffmpeg", "-y", "-i", input.VideoPath, "-vf", "scale="+strconv.Itoa(dimensions.Width)+":"+strconv.Itoa(dimensions.Height), "-c:a", "copy", input.OutputPath)

	_, err := util.RunCommand(cmd)

	if err != nil {
		return err
	}

	return nil

}
